proc_macros/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+        let mut variants: Option<Vec<Ident>> = Default::default();
+        while !input.is_empty() {
+            let key: Ident = input.parse()?;
+            input.parse::<Token![=>]>()?;
+            match &*key.to_string() {
+                "name" => {
+                    name = Some(input.parse::<Ident>()?);
+                }
+                "variants" => {
+                    variants = Some(parse_idents_array(input)?);
+                }
+                _ => panic!("didn't expect key {}", key),
+            }
+            input.parse::<Token![,]>()?;
+        }
+        Ok(FixedMapArgs {
+            name: name.expect("Expected name specifier"),
+            variants: variants.expect("Expected variants specifier"),
+        })
+    }
+}
+
+#[proc_macro]
+pub fn fixed_map(input: TokenStream) -> TokenStream {
+    let FixedMapArgs { name, variants } = parse_macro_input!(input as FixedMapArgs);
+
+    let collection_type_name = format_ident!("By{name}");
+    let all_variants_collection_name = format_ident!(
+        "ALL_{}",
+        name.to_string().to_plural().to_screaming_snake_case()
+    );
+
+    let token_enum_definition = get_token_enum_definition(&name, &variants);
+    let collection_type_definition =
+        get_collection_type_definition(&collection_type_name, &variants);
+    let iter_implementations =
+        get_iter_implementations(&name, &collection_type_name, &all_variants_collection_name);
+    let index_implementations = get_index_implementations(&name, &collection_type_name);
+    let deref_implementation = get_deref_implementation(&collection_type_name, &variants);
+    let from_usize_implementation = get_from_usize_implementation(&name, &variants);
+    let all_variants_collection_definition = get_all_variants_collection_definition(
+        &name,
+        &collection_type_name,
+        &variants,
+        &all_variants_collection_name,
+    );
+    let collection_generator_macro_definition =
+        get_collection_generator_macro_definition(&collection_type_name, &variants);
+
+    quote! {
+        #token_enum_definition
+
+        #collection_type_definition
+
+        #iter_implementations
+
+        #index_implementations
+
+        #deref_implementation
+
+        #from_usize_implementation
+
+        #all_variants_collection_definition
+
+        #collection_generator_macro_definition
+    }
+    .into()
+}
+
+fn get_token_enum_definition(name: &Ident, variants: &[Ident]) -> proc_macro2::TokenStream {
+    quote! {
+        #[derive(Copy, Clone, Debug, Eq, PartialEq, clap::ValueEnum)]
+        pub enum #name {
+            #(#variants),*
+        }
+    }
+}
+
+fn get_collection_type_definition(
+    collection_type_name: &Ident,
+    variants: &[Ident],
+) -> proc_macro2::TokenStream {
+    let len = variants.len();
+    quote! {
+        #[derive(Default)]
+        pub struct #collection_type_name<T>([T; #len]);
+    }
+}
+
+fn get_iter_implementations(
```

Comment:
Per above comment this is implementing "`HashMap`-style" `.iter()`/`.values()` on eg `BySupportedLanguage`